File: vedha_ai/app/schemas/ai.py
"""
app/schemas/ai.py
Pydantic v2 schemas for all AI API endpoints.
Covers: Career Mentor, Salary, Career Path, Trends,
        Resume Builder, Skill Gap, ATS,
        Coding Assistant, Research, PDF Chat.
"""
from typing import Any
from pydantic import BaseModel, Field


# ──────────────────────────────────────────────
# Career Mentor
# ──────────────────────────────────────────────

class CareerMentorRequest(BaseModel):
    message: str = Field(min_length=2, max_length=2000)
    current_role: str = Field(default="Student", max_length=100)
    target_role: str = Field(default="Software Engineer", max_length=100)
    skills: list[str] = Field(default=[])

class CareerMentorResponse(BaseModel):
    reply: str
    session_active: bool
    status: str

class ClearChatRequest(BaseModel):
    context: str = Field(default="career", max_length=50)


# ──────────────────────────────────────────────
# Salary Prediction
# ──────────────────────────────────────────────

class SalaryPredictionRequest(BaseModel):
    role: str = Field(min_length=2, max_length=100)
    skills: list[str] = Field(default=[])
    experience_years: int = Field(default=0, ge=0, le=50)
    location: str = Field(default="India", max_length=100)

class SalaryRangeItem(BaseModel):
    min: float
    max: float
    currency: str = "INR"
    period: str = "annual"

class SalaryPredictionResponse(BaseModel):
    role: str
    location: str
    predicted_lpa: float | None = None
    entry_level: SalaryRangeItem | None = None
    mid_level: SalaryRangeItem | None = None
    senior_level: SalaryRangeItem | None = None
    growth_trajectory: str | None = None
    top_paying_companies: list[str] = []
    reasoning: str | None = None
    status: str


# ──────────────────────────────────────────────
# Career Path Prediction
# ──────────────────────────────────────────────

class CareerPathRequest(BaseModel):
    current_role: str = Field(default="Student", max_length=100)
    experience_years: int = Field(default=0, ge=0, le=50)
    skills: list[str] = Field(default=[])
    interests: list[str] = Field(default=[])

class PredictedRole(BaseModel):
    role: str
    timeline: str
    probability: float = Field(..., ge=0.0, le=1.0)
    required_skills: list[str]

class CareerPathResponse(BaseModel):
    predicted_roles: list[PredictedRole] = []
    recommended_path: str | None = None
    skill_investments: list[str] = []
    market_demand: str | None = None
    growth_probability: float | None = Field(default=None, ge=0.0, le=1.0)
    industry_outlook: str | None = None
    action_plan: list[str] = []
    status: str


# ──────────────────────────────────────────────
# Industry Trends
# ──────────────────────────────────────────────

class TrendsRequest(BaseModel):
    domain: str = Field(min_length=2, max_length=100)
    skills: list[str] = Field(default=[])

class TrendsResponse(BaseModel):
    trending_skills: list[str] = []
    declining_skills: list[str] = []
    emerging_technologies: list[str] = []
    top_companies_hiring: list[str] = []
    average_salary_trend: str | None = None
    demand_index: int | None = None
    supply_index: int | None = None
    outlook: str | None = None
    key_insights: list[str] = []
    status: str


# ──────────────────────────────────────────────
# Resume Builder
# ──────────────────────────────────────────────

class ResumeBuilderRequest(BaseModel):
    target_role: str = Field(min_length=2, max_length=100)
    experience: str = Field(min_length=10, max_length=5000)
    education: str = Field(min_length=10, max_length=2000)
    skills: list[str] = Field(min_length=1)
    projects: list[str] = Field(default=[])

class ResumeBuilderResponse(BaseModel):
    target_role: str
    summary: str | None = None
    skills_section: list[str] = []
    experience_bullets: list[str] = []
    project_highlights: list[str] = []
    ats_keywords: list[str] = []
    formatting_tips: list[str] = []
    estimated_ats_score: int | None = None
    status: str


# ──────────────────────────────────────────────
# Skill Gap Analysis
# ──────────────────────────────────────────────

class SkillGapRequest(BaseModel):
    target_role: str = Field(min_length=2, max_length=100)
    current_skills: list[str] = Field(default=[])
    matched_skills: list[str] = Field(default=[])
    missing_skills: list[str] = Field(default=[])

class LearningPlanItem(BaseModel):
    skill: str
    priority: str
    estimated_weeks: int
    resources: list[str]
    reason: str

class SkillGapResponse(BaseModel):
    target_role: str
    gap_score: int | None = None
    critical_missing: list[str] = []
    nice_to_have: list[str] = []
    learning_plan: list[LearningPlanItem] = []
    total_weeks_to_ready: int | None = None
    summary: str | None = None
    status: str


# ──────────────────────────────────────────────
# ATS Analysis
# ──────────────────────────────────────────────

class ATSAnalysisRequest(BaseModel):
    resume_text: str = Field(min_length=100, max_length=20000)
    target_role: str = Field(min_length=2, max_length=100)
    ats_score: int = Field(ge=0, le=100)
    matched_skills: list[str] = Field(default=[])
    missing_skills: list[str] = Field(default=[])

class ATSAnalysisResponse(BaseModel):
    ats_score: int
    sections: dict[str, Any] = {}
    keyword_density: float | None = None
    formatting_issues: list[str] = []
    missing_keywords: list[str] = []
    improvement_actions: list[str] = []
    overall_grade: str | None = None
    status: str


# ──────────────────────────────────────────────
# Coding Assistant
# ──────────────────────────────────────────────

class CodeExplainRequest(BaseModel):
    code: str = Field(min_length=1, max_length=10000)
    language: str = Field(default="python", max_length=30)

class CodeOptimizeRequest(BaseModel):
    code: str = Field(min_length=1, max_length=10000)
    language: str = Field(default="python", max_length=30)

class CodeDebugRequest(BaseModel):
    code: str = Field(min_length=1, max_length=10000)
    error: str = Field(min_length=1, max_length=2000)
    language: str = Field(default="python", max_length=30)

class BigORequest(BaseModel):
    code: str = Field(min_length=1, max_length=10000)
    language: str = Field(default="python", max_length=30)

class CodingHintRequest(BaseModel):
    problem_title: str = Field(min_length=2, max_length=200)
    problem_desc: str = Field(min_length=10, max_length=5000)
    current_code: str = Field(default="", max_length=10000)
    language: str = Field(default="python", max_length=30)

class CodingAssistantRequest(BaseModel):
    code: str = Field(min_length=1, max_length=10000)
    problem_id: int = Field(ge=1)
    language: str = Field(default="python", max_length=30)
    error_log: str | None = Field(default=None, max_length=2000)

class CodingAssistantResponse(BaseModel):
    time_complexity: str | None = None
    space_complexity: str | None = None
    explanation: str | None = None
    debug_hint: str | None = None
    suggested_fix: str | None = None
    status: str


# ──────────────────────────────────────────────
# Research Assistant
# ──────────────────────────────────────────────

class ResearchRequest(BaseModel):
    query: str = Field(min_length=5, max_length=1000)

class ResearchResponse(BaseModel):
    query: str
    answer: str
    sources: list[str] = []
    status: str


# ──────────────────────────────────────────────
# PDF Chat
# ──────────────────────────────────────────────

class PDFChatRequest(BaseModel):
    question: str = Field(min_length=5, max_length=1000)
    session_id: str = Field(default="pdf", max_length=100)

class PDFChatResponse(BaseModel):
    question: str
    answer: str
    session_id: str
    status: str

File: vedha_ai/vedha-frontend/src/ai/services/ai.ts
import { api } from "@/services/api";
import type {
  ChatMessage,
  CareerMentorResponse,
  ResumeAnalysisResponse,
  SkillGapResponse,
  RoadmapResponse,
  CodingHelpResponse,
  MockInterviewResponse,
  JobMatchResponse,
  CareerPredictionResponse,
  ResearchAssistantResponse,
  PdfChatResponse,
} from "../types";

// Dynamic availability detection state
let backendOnline = false;

export async function checkBackendAvailability(): Promise<boolean> {
  try {
    const res = await api.get("/health", { timeout: 1500 });
    backendOnline = res.status === 200;
  } catch {
    backendOnline = false;
  }
  return backendOnline;
}

// Auto-trigger backend availability check
checkBackendAvailability();

export function isBackendOnline(): boolean {
  return backendOnline;
}

// 1. AI Career Mentor
export async function chatWithCareerMentor(
  message: string,
  _history: ChatMessage[]
): Promise<CareerMentorResponse> {
  if (isBackendOnline()) {
    const res = await api.post<{ reply: string; session_active: boolean; status: string }>("/ai/mentor/chat", {
      message,
      current_role: "Student",
      target_role: "Software Engineer",
      skills: [],
    });
    return {
      reply: res.data.reply,
      predictedRoles: [],
    };
  }
  // Mock fallback
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve({
        reply: `That is an excellent goal! Deepening your knowledge in database connectivity pools like SQL and PostgreSQL will make you highly competitive. I recommend focusing on FastAPI's dependency yield yield pattern.`,
        predictedRoles: ["Backend Engineer", "DevOps Practitioner"],
      });
    }, 800);
  });
}

// 2. AI Resume Analyzer & Builder & ATS score checks
export async function analyzeResume(
  resumeFileContent: string,
  targetRole: string
): Promise<ResumeAnalysisResponse> {
  if (isBackendOnline()) {
    const res = await api.post<ResumeAnalysisResponse>("/ai/resume/analyze", {
      resumeFileContent,
      targetRole,
    });
    return res.data;
  }
  // Mock fallback
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve({
        atsScore: 89,
        extractedSkills: ["React", "TypeScript", "JavaScript", "HTML/CSS"],
        missingSkills: ["Docker", "FastAPI", "PostgreSQL"],
        feedback: [
          {
            section: "Impact Metrics",
            score: 75,
            suggestions: ["Use more action verbs", "Quantify project metrics by adding percentages (e.g. Optimized queries by 35%)."],
          },
          {
            section: "ATS Compatibility",
            score: 95,
            suggestions: ["Layout is clean and standard single-column format."],
          },
        ],
      });
    }, 1000);
  });
}

// 3. AI Skill Gap Analysis
export async function analyzeSkillGap(
  currentSkills: string[],
  targetRole: string
): Promise<SkillGapResponse> {
  if (isBackendOnline()) {
    const res = await api.post<any>("/ai/resume/skill-gap", {
      target_role: targetRole,
      current_skills: currentSkills,
      matched_skills: [],
      missing_skills: []
    });
    const data = res.data;
    return {
      matchPercentage: 100 - (data.gap_score ?? 30),
      presentSkills: currentSkills.map((s) => ({ name: s, proficiency: 85 })),
      missingSkills: [
        ...data.critical_missing.map((s: string) => ({ name: s, priority: "High" as const })),
        ...data.nice_to_have.map((s: string) => ({ name: s, priority: "Medium" as const }))
      ]
    };
  }
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve({
        matchPercentage: 70,
        presentSkills: currentSkills.map((s) => ({ name: s, proficiency: 85 })),
        missingSkills: [
          { name: "FastAPI", priority: "High" },
          { name: "Docker", priority: "High" },
          { name: "SQL PostgreSQL", priority: "Medium" },
        ],
      });
    }, 600);
  });
}

// 4. AI Learning Roadmap
export async function getAILearningRoadmap(targetRole: string): Promise<RoadmapResponse> {
  if (isBackendOnline()) {
    const res = await api.get<RoadmapResponse>(`/ai/roadmap?role=${encodeURIComponent(targetRole)}`);
    return res.data;
  }
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve({
        targetRole,
        completion: 60,
        nodes: [
          { id: 1, name: "Web Basics (HTML/CSS)", desc: "Build foundational static interfaces.", status: "completed", estimatedHours: 12 },
          { id: 2, name: "JavaScript Advanced & Async", desc: "Master asynchronous code execution, callbacks, and promises.", status: "completed", estimatedHours: 18 },
          { id: 3, name: "React Web Applications", desc: "Build modular interfaces, component lifecycles, states, and hooks.", status: "in_progress", estimatedHours: 24 },
          { id: 4, name: "Backend Rest API Design", desc: "Configure FastAPI routes, database integrations, and auth schemas.", status: "locked", estimatedHours: 20 },
          { id: 5, name: "Cloud Containers (Docker/K8s)", desc: "Scale and deploy containers onto AWS.", status: "locked", estimatedHours: 15 },
        ],
      });
    }, 700);
  });
}

// 5. AI Coding Assistant
export async function getCodingAssistantHelp(
  code: string,
  problemId: number,
  language: string,
  errorLog?: string
): Promise<CodingHelpResponse> {
  if (isBackendOnline()) {
    const res = await api.post<CodingHelpResponse>("/ai/coding/assistant", {
      code,
      problemId,
      language,
      errorLog,
    });
    return res.data;
  }
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve({
        debugHint: "Ensure you handle index boundaries correctly. Your current loop might result in an IndexOutOfBounds exception when the array is empty.",
        explanation: "This algorithm targets finding two elements summing up to a target value. A hashmap yields O(n) runtime complexity instead of nesting double loops.",
        suggestedFix: `// Optimization using HashMap:\nconst map = new Map();\nfor (let i = 0; i < nums.length; i++) {\n  const complement = target - nums[i];\n  if (map.has(complement)) return [map.get(complement), i];\n  map.set(nums[i], i);\n}`,
        timeComplexity: "O(n)",
        spaceComplexity: "O(n)",
      });
    }, 900);
  });
}

// 6. AI Mock Interview & Feedback
export async function conductMockInterview(
  interviewId: number,
  transcript: { question: string; answer: string }[]
): Promise<MockInterviewResponse> {
  if (isBackendOnline()) {
    const res = await api.post<MockInterviewResponse>("/ai/interview/simulate", {
      interviewId,
      transcript,
    });
    return res.data;
  }
  return new Promise((resolve) => {
    setTimeout(() => {
      const questionIndex = transcript.length;
      const questions = [
        "What is the difference between client-side rendering and server-side rendering?",
        "Explain how React Virtual DOM updates UI changes efficiently.",
        "How do you manage race conditions in asynchronous actions in Redux or React Hooks?",
      ];
      if (questionIndex >= questions.length) {
        resolve({
          nextQuestion: "Interview completed.",
          isFinished: true,
          overallScore: 84,
          transcripts: transcript.map((t, idx) => ({
            ...t,
            score: 80 + idx * 5,
            feedback: "Answer is clear. Incorporating actual project metrics would enhance authenticity.",
          })),
        });
      } else {
        resolve({
          nextQuestion: questions[questionIndex],
          isFinished: false,
        });
      }
    }, 800);
  });
}

// 7. AI Job Matching
export async function matchJobSuitability(
  studentSkills: string[],
  jobDescription: string
): Promise<JobMatchResponse> {
  if (isBackendOnline()) {
    const res = await api.post<any>("/ai/resume/job-match", {
      student_skills: studentSkills,
      job_description: jobDescription,
    });
    return {
      suitabilityScore: res.data.suitabilityScore,
      alignmentReasoning: res.data.alignmentReasoning,
      suggestedAction: res.data.suggestedAction,
    };
  }
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve({
        suitabilityScore: 85,
        alignmentReasoning: "Your skill profile overlaps strongly with React, TypeScript, and Git requirements. Lack of Docker is the only gap.",
        suggestedAction: "Complete the Docker Container milestone inside your roadmap to boost compatibility to 95%.",
      });
    }, 700);
  });
}

// 8. AI Career & Salary Prediction & Industry Trends
export async function predictCareerPath(): Promise<CareerPredictionResponse> {
  if (isBackendOnline()) {
    try {
      const [profileRes, skillsRes] = await Promise.all([
        api.get<any>("/profile/me").catch(() => ({ data: null })),
        api.get<any[]>("/skill/").catch(() => ({ data: [] })),
      ]);

      const skills = skillsRes.data?.map((s: any) => s.skill_name) || [];
      const currentRole = profileRes.data?.target_role || "Student";
      const experienceYears = parseInt(profileRes.data?.experience) || 0;
      const location = profileRes.data?.location || "India";

      const [careerRes, salaryRes, trendsRes] = await Promise.all([
        api.post<any>("/ai/mentor/career-path", {
          current_role: currentRole,
          experience_years: experienceYears,
          skills: skills,
          interests: [],
        }),
        api.post<any>("/ai/mentor/salary", {
          role: currentRole,
          skills: skills,
          experience_years: experienceYears,
          location: location,
        }),
        api.post<any>("/ai/mentor/trends", {
          domain: "AI/ML",
          skills: skills,
        }),
      ]);

      const careerData = careerRes.data;
      const salaryData = salaryRes.data;
      const trendsData = trendsRes.data;

      return {
        careerRoadmap: careerData.predicted_roles?.map((r: any) => `${r.role} (${r.timeline})`) || [currentRole],
        growthProbability: Math.round((careerData.growth_probability ?? 0.8) * 100),
        salaryProjection: [
          {
            region: `${salaryData.location || location} (LPA / k Currency)`,
            values: [
              { year: "Entry Level", lpa: salaryData.entry_level?.max || 6 },
              { year: "Mid Level", lpa: salaryData.mid_level?.max || 12 },
              { year: "Senior Level", lpa: salaryData.senior_level?.max || 20 },
            ],
          },
        ],
        supplyDemandIndex: trendsData.trending_skills?.map((skill: string) => ({
          skill,
          supply: Math.floor(Math.random() * 30) + 10,
          demand: Math.floor(Math.random() * 40) + 60,
        })) || [],
      };
    } catch (error) {
      console.error("Failed to query career predictions from backend, falling back to mocks", error);
    }
  }
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve({
        careerRoadmap: ["Junior Frontend Developer", "Full Stack Architect", "Technical Director"],
        growthProbability: 92,
        salaryProjection: [
          {
            region: "India (LPA INR)",
            values: [
              { year: "Year 1", lpa: 8 },
              { year: "Year 2", lpa: 11 },
              { year: "Year 3", lpa: 15 },
              { year: "Year 4", lpa: 21 },
            ],
          },
        ],
        supplyDemandIndex: [
          { skill: "FastAPI", supply: 30, demand: 85 },
          { skill: "React", supply: 90, demand: 95 },
          { skill: "Docker", supply: 25, demand: 75 },
        ],
      });
    }, 800);
  });
}

// 9. AI Research Assistant
export async function queryResearchPapers(query: string): Promise<ResearchAssistantResponse> {
  if (isBackendOnline()) {
    const res = await api.post<ResearchAssistantResponse>("/ai/research/query", { query });
    return res.data;
  }
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve({
        papers: [
          {
            id: "paper-1",
            title: "Attention Is All You Need",
            authors: "Vaswani et al.",
            year: "2017",
            abstract: "We propose a new simple network architecture, the Transformer, based solely on attention mechanisms, dispensing with recurrence and convolutions entirely.",
            citationCount: 112000,
          },
          {
            id: "paper-2",
            title: "BERT: Pre-training of Deep Bidirectional Transformers",
            authors: "Devlin et al.",
            year: "2018",
            abstract: "We introduce a new language representation model called BERT, which stands for Bidirectional Encoder Representations from Transformers.",
            citationCount: 45000,
          },
        ],
        summaryMarkdown: "### Key Takeaways:\n- Transformers replaced RNNs/LSTMs in sequence tasks.\n- Multi-head self-attention enables parallelized processing of long contexts.",
      });
    }, 1000);
  });
}

// 10. AI PDF Chat
export async function chatWithPdf(
  _fileId: string,
  question: string
): Promise<PdfChatResponse> {
  if (isBackendOnline()) {
    const res = await api.post<any>("/ai/mentor/pdf/chat", {
      question,
      session_id: "pdf",
    });
    return {
      answer: res.data.answer,
      citations: [],
    };
  }
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve({
        answer: "Based on the uploaded syllabus document, Section 4 details FastAPI route parameters setups, yield dependencies setup, and PostgreSQL connection pool sizing.",
        citations: [
          { filename: "FastAPI_Syllabus.pdf", page: 4, snippet: "The db session dependency is initialized using yield session context managers..." },
        ],
      });
    }, 900);
  });
}

// 11. PDF Document Upload
export async function uploadPdf(file: File): Promise<any> {
  const formData = new FormData();
  formData.append("file", file);
  const response = await api.post("/ai/mentor/pdf/upload", formData, {
    headers: {
      "Content-Type": "multipart/form-data",
    },
  });
  return response.data;
}

